File: exercises/advent2016/problem5.rb
require 'digest'

def md5(str)
  Digest::MD5.hexdigest(str)
end

#puts md5("abc3231929").start_with?("00000")

i = 0
result = []
while result.size < 8
  s = md5("reyedfim#{i}")
  if s.start_with?("00000")
    result << s[5]
  end
  i += 1
end

puts result.join

File: exercises/advent2016/problem5.clj
(ns problem5)

(def doorid "abc")

(defn md5 [s]
  (apply str
         (map (partial format "%02x")
              (.digest (doto (java.security.MessageDigest/getInstance "MD5")
                         .reset
                         (.update (.getBytes s)))))))

(defn start-with-zeros? [s]
  (clojure.string/starts-with? s "00000"))

(defn sixth-char [s]
  (nth s 5))

(println "Running")
(println
 (->> (iterate inc 0)
      (map (partial str doorid))
      (map md5)
      (filter start-with-zeros?)
      (map sixth-char)
      (take 2)
      (clojure.string/join "")))

